### ISSUE TITLE
How to set custom shortcuts

### ISSUE THREAD

**Issue by spiderbit** [atlas-engineer/next#447]
I want to set the following keybindings:

(define-key :keymap *my-keymap* "r" #'history-forwards)
(define-key :keymap *my-keymap* "g" #'history-backwards)
(define-key :keymap *my-keymap* "u" #'set-url-current-buffer)
(define-key :keymap *my-keymap* "c" #'scroll-up)
(define-key :keymap *my-keymap* "t" #'scroll-down)
(define-key :keymap *my-keymap* "d" #'scroll-page-up)
(define-key :keymap *my-keymap* "s" #'scroll-page-down) 

It's inspired by https://github.com/xahlee/xah-fly-keys Dvorak layout and what I currently use for my browsers. (vimium extension for firefox/chromium, vimb...)

**Comment by Ambrevar** [comment]
Hmm... What's your question? :p 
What you wrote is correct, then you need to actually use the keymap in a mode, like this:

```lisp
(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-schemes :initform (list :emacs-map *my-keymap*
                                   :vi-normal *my-keymap*))))

(defun my-buffer-defaults (buffer)
  (dolist (mode '(proxy-mode my-mode)) ;; <= Any mode that you want by default!
    (pushnew mode (default-modes buffer))))

(defun my-interface-defaults ()
  (hooks:add-to-hook (hooks:object-hook *interface* 'buffer-make-hook)
                     #'my-buffer-defaults))
```

Hope that helps!

**Comment by spiderbit** [comment]
I get the following error from this line:


>  The variable NEXT::MY-KEYMAP is unbound.


is there some (make-keymap) ? If yes, it would be good to add that to the documentation.

**Comment by Ambrevar** [comment]
Indeed you can create a keymap as follows:

```lisp
(defvar *my-keymap* (make-keymap)
  "My keymap.")
```

> If yes, it would be good to add that to the documentation.

Will do, thanks for the suggestion!

**Comment by spiderbit** [comment]
Well thanks now it don't throws a error but it doesn't do what it supposed to do, none of the shortcuts works.

And it doesn't recognize the 2 history functions even in the console they are a choice.

> The function NEXT::HISTORY-FORWARDS is undefined.

Do I have to write wrapper functions around that functions which is what I see in the extensions.org documentation file

Ideally you could post a minimal init file that works on your side? with that shortcuts? Before we go another 10 times forth and back?

**Comment by tviti** [comment]
I get the same error as you when I try to add the same keybinding to my keybinding Looks like `history-forwards` is defined in the `web-mode` package, and isn't actually visible from `:next` (I guess not till the buffer gets created, and `web-mode` is activated)?

Try this:
```
(define-key :keymap my-keymap "r" #'next/web-mode:history-forwards)
```

**Comment by Ambrevar** [comment]
Indeed, the history commands belong to web mode.

We should display it in the `execute-command' minibuffer.

Only bindings of active modes are caught.  So if web-mode is not active,
the history-forwards command won't be available either.

Hope that helps! :)

**Comment by spiderbit** [comment]
Well now history function works btw here on github always the star "*" get suppressed around my-keymap apperently sucks when you poste code and it hides parts of it, don't get how that can happen...

But the other commands seem to work, too. But history is buggy should I open a different bugreport for that?

1. after opening a url back brings me not back to the start page. Even if I click on a link on the startpage back don't brings me back. So the startpage is not in the history can be seen as feature not sure feels a bit odd to me.
2. very strange behaviour that is different on differenent sides: so if I go to google or github and I press history back it reloads the site and does the resizing I set up in the hook (hidpi) which is odd if it has only 1 site in the history forward and backward should not reload the page it should just do nothing right?
But when I open heise.de (a german IT news site) it does ignore history back/forth and not reloads the site?

Should I open a new bug for that?

**Comment by Ambrevar** [comment]
Stefan Huchler <notifications@github.com> writes:

> Well now history function works btw here on github always the star "*"
> get suppressed around my-keymap apperently sucks when you poste code
> and it hides parts of it, don't get how that can happen...

Sorry, didn't get that.  Can you rephrase?

> But the other commands seem to work, too. But history is buggy should I open a different bugreport for that?
>
> 1. after opening a url back brings me not back to the start page. Even
> if I click on a link on the startpage back don't brings me back. So
> the startpage is not in the history can be seen as feature not sure
> feels a bit odd to me.

Indeed, the start page is not in the history.  It's intentional, but I
see no reason it shouldn't, so I guess we can call this a bug and we
should fix this.


> 2. very strange behaviour that is different on differenent sides: so
> if I go to google or github and I press history back it reloads the
> site and does the resizing I set up in the hook (hidpi) which is odd
> if it has only 1 site in the history forward and backward should not
> reload the page it should just do nothing right?

Another bug! :p  Thanks for catching this!

The buffer history implementation was just rewritten, so it's a bit
rough on the edge, sorry about that.

> But when I open heise.de (a german IT news site) it does ignore
> history back/forth and not reloads the site?

I can't reproduce.  Can you provide a precise recipe (in a new issue)?

> Should I open a new bug for that?

Please do! :)

**Comment by spiderbit** [comment]
K I checked again the handling of the star/multiply sign.

In my original post I posted the source as normal text not as source therefor it removed the * around the my-keymap strings when I saw your code and that I had to add the * around it to make it work:

`(define-key :keymap my-keymap "r" #'next/web-mode:history-forwards)`

I assumed you also did have the * around the my-keymap and it got stripped from it, if you mark it as source it don't read it as markup, you just forgot them.
So that is fine.



About the heise.de site besides the point that the startpage is not in the history it works fine it's just the google/github sites that don't work correctly so it makes no sense to report a bug because 1 site works correct I think?

It's just odd to me why it site specific works differently...

**Comment by Ambrevar** [comment]
Yes, please open a new issue and list all the problems you've noticed
with the history, with precise recipes if possible.

Regarding the `*my-keymap*` with or without stars: the stars are purely
conventional, they are part of the symbol name like any other character.
They are use in Common Lisp to denote global variable.

You don't have to use them, but you've got to be consistent with your
use of the name throughout the configuration file.

**Comment by azzamsa** [comment]
Is this the correct MWE of overriding the default keybinding?


```cl
(in-package :next)

(defvar *my-keymap* (make-keymap)
  "My keymap.")

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-schemes :initform (list :emacs-map *my-keymap*
                                   :vi-normal *my-keymap*))))

(defun my-buffer-defaults (buffer)
  (dolist (mode '(proxy-mode my-mode)) ;; <= Any mode that you want by default!
    (pushnew mode (default-modes buffer))))

(defun my-interface-defaults ()
  (hooks:add-to-hook (hooks:object-hook *interface* 'buffer-make-hook)
                     #'my-buffer-defaults))

(define-key :keymap *my-keymap* "s-m" #'switch-buffer)
```

1) start next
2) do `M-x my-mode`
3) press `s-m` (windows + m). But nothing happened. I've tried to use different key but didn't work either.

Did I miss something?

**Comment by Ambrevar** [comment]
You are not installing the new interface defaults ;)

You just miss this line:

```
(hooks:add-to-hook '*after-init-hook* #'my-interface-defaults)
```

By the way, on master we've got a new (experimental) configuration
style, the CLOS-style config.

The `my-mode` keymap is mostly useful to create _global_ bindings and
override other bindings.  If you want to customize the bindings of a
specific mode, you can create a subclass of the mode, then use it
instead of the original.

He follow the CLOS-style way to add my-mode to the list of mode for all
buffers:


```
(defclass my-buffer (buffer)
  ((default-modes :initform (append
                             '(my-mode vi-normal-mode
                               my-blocker-mode
                               proxy-mode)
                             (get-default 'buffer 'default-modes)))
   (load-hook :initform (list #'old-reddit-handler
                              #'auto-proxy-handler))))

(setf *buffer-class* 'my-buffer)
```

In a similar fashion, you can create a new mode with new bindings and
call

```
(setf *WHATEVER-MODE-class* 'my-WHATEVER-MODE)
```

(Of course, you need to replace WHATEVER-MODE with what is appropriate).

**Comment by azzamsa** [comment]
> You are not installing the new interface defaults ;) You just miss this line:

Where to add that line? is it here?

``` diff
+ (hooks:add-to-hook '*after-init-hook* #'my-interface-defaults)
(define-key :keymap *my-keymap* "C-x m" #'switch-buffer)
```

If yes, it didn't manage to work either. I've tried the key to "C-x m", "s-m", "m". Am I missing something?

**Comment by Ambrevar** [comment]
Sorry, I didn't spot the typo at first: it's `:emacs`, not `:emacs-map`.
I've fixed the documentation accordingly.

To sum up, your init.lisp:

```
(in-package :next-user)

(defvar *my-keymap* (make-keymap)
  "My keymap.")

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-schemes :initform (list :emacs *my-keymap*
                                   :vi-normal *my-keymap*))))

(defun my-buffer-defaults (buffer)
  (dolist (mode '(proxy-mode my-mode)) ;; <= Any mode that you want by default!
    (pushnew mode (default-modes buffer))))

(defun my-interface-defaults ()
  (hooks:add-to-hook (hooks:object-hook *interface* 'buffer-make-hook)
                     #'my-buffer-defaults))

(define-key :keymap *my-keymap* "s-m" #'switch-buffer)
(hooks:add-to-hook '*after-init-hook* #'my-interface-defaults)
```

It'd be nice to spot this kind of errors.
Maybe declare a list of known keymap schemes globally?

**Comment by azzamsa** [comment]
> Maybe declare a list of known keymap schemes globally?

Your latest snippets works for me. So you changed `(in-package :next-user)` and `(list :emacs *my-keymap*`.

Thank you :)

**Comment by Ambrevar** [comment]
The :next-user package is not mandatory, but recommended.

Your actualy issue was due to a typo in the doc: it's `:emacs`, not `:emacs-map`.

**Comment by azzamsa** [comment]
Okay thanks, everything sets. I am unable to use `super` key, and want to open new issue, turns out it was `S` [on next](https://github.com/atlas-engineer/next/blob/master/documents/MANUAL.org#keybinding) (in emacs it's `s`).

Thank you.

**Comment by jmercouris** [comment]
If the problem is solved, may we close this issue?

**Comment by azzamsa** [comment]
@jmercouris I am not the author, :)

**Comment by Ambrevar** [comment]
I think we can close this now.  Feel free to reopen if there is anything else.

**Comment by dalanicolai** [comment]
Hi! I am trying to set custom keybindings for vi-mode. However the code in the documentation is not clear to me/does not work. For example if I use the init file as given in [this comment above](https://github.com/atlas-engineer/next/issues/447#issuecomment-562519272) then I get the following error message:

```
Next version 1.5.0-73ab116f
Loading Lisp file #P"/home/dalanicolai/.config/next/init.lisp"...
While evaluating the form starting at line 3, column 0
  of #P"/home/dalanicolai/.config/next/init.lisp":
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {10006F05B3}>:
  Error: we could not load the init file:
invalid number of arguments: 0

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10006F05B3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "~a:~&~a" {1006389C63}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "~a:~&~a" {1006389C63}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "~a:~&~a" {1006389C63}>)
3: (UIOP/IMAGE:HANDLE-FATAL-CONDITION #<SIMPLE-ERROR "~a:~&~a" {1006389C63}>)
4: (SB-KERNEL::%SIGNAL #<SIMPLE-ERROR "~a:~&~a" {1006389C63}>)
5: (ERROR "~a:~&~a" "Error: we could not load the init file" #<SB-INT:SIMPLE-PROGRAM-ERROR "invalid number of arguments: ~S" {1006389813}>)
6: (NEXT::LOAD-LISP #P"/home/dalanicolai/.config/next/init.lisp" :INTERACTIVE T)
7: (NEXT:START :URLS NIL :INIT-FILE NIL)
8: ((LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE))
9: (UIOP/IMAGE:CALL-WITH-FATAL-CONDITION-HANDLER #<CLOSURE (LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE) {100621BE9B}>)
10: ((FLET SB-UNIX::BODY :IN SAVE-LISP-AND-DIE))
11: ((FLET "WITHOUT-INTERRUPTS-BODY-14" :IN SAVE-LISP-AND-DIE))
12: ((LABELS SB-IMPL::RESTART-LISP :IN SAVE-LISP-AND-DIE))

unhandled condition in --disable-debugger mode, quitting
```
line 3 is the `defvar` form

Is this a bug? Thanks for your help... (and the beautiful browser!)

**Comment by Ambrevar** [comment]
Thank you for reporting, there was a documentation bug indeed!

Fixed in commit 14bf6c4801c8541871aa7735f38e4e2cf502f593.

- `make-keymap` takes a name (a string) as argument.
- I recommend you use `define-configuration` to simplify the configuration.

Minimal example:

```lisp
(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-scheme :initform
                  (define-scheme "my-scheme"
                    scheme:cua
                    (list
                     "C-f" 'history-forwards
                     "C-b" 'history-backwards)

                    scheme:emacs
                    (list ...)))))

(define-configuration buffer
  ((default-modes (append '(my-mode) %slot-default))))
```

**Comment by dalanicolai** [comment]
Thanks for looking into it. Unfortunately I do not understand much yet about common-lisp or from the next documentation.

When I try the new code from the documentation, for example the following part
```
(defvar *my-keymap* (make-keymap "my-map")
(define-key *my-keymap*
  "C-f" 'history-forwards
  "C-b" 'history-backwards)

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-scheme :initform (keymap:make-scheme
                             scheme:emacs *my-keymap*
                             scheme:vi-normal *my-keymap*))))

(define-configuration buffer
  ((default-modes (append '(my-mode) %slot-default))))
```
then I still get an error:
```
Next version 1.5.0-73ab116f
Loading Lisp file #P"/home/dalanicolai/.config/next/init.lisp"...
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {10006F05B3}>:
  Error: we could not load the init file:
READ error during LOAD:

  end of file on #<SB-INT:FORM-TRACKING-STREAM for "file /home/dalanicolai/.config/next/init.lisp" {1006373313}>

  (in form starting at line: 1, column: 0, position: 0)

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10006F05B3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "~a:~&~a" {1006376523}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "~a:~&~a" {1006376523}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "~a:~&~a" {1006376523}>)
3: (UIOP/IMAGE:HANDLE-FATAL-CONDITION #<SIMPLE-ERROR "~a:~&~a" {1006376523}>)
4: (SB-KERNEL::%SIGNAL #<SIMPLE-ERROR "~a:~&~a" {1006376523}>)
5: (ERROR "~a:~&~a" "Error: we could not load the init file" #<SB-C::INPUT-ERROR-IN-LOAD {1006376353}>)
6: (NEXT::LOAD-LISP #P"/home/dalanicolai/.config/next/init.lisp" :INTERACTIVE T)
7: (NEXT:START :URLS NIL :INIT-FILE NIL)
8: ((LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE))
9: (UIOP/IMAGE:CALL-WITH-FATAL-CONDITION-HANDLER #<CLOSURE (LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE) {100621BE9B}>)
10: ((FLET SB-UNIX::BODY :IN SAVE-LISP-AND-DIE))
11: ((FLET "WITHOUT-INTERRUPTS-BODY-14" :IN SAVE-LISP-AND-DIE))
12: ((LABELS SB-IMPL::RESTART-LISP :IN SAVE-LISP-AND-DIE))

unhandled condition in --disable-debugger mode, quitting
```
I guess maybe the `defvar` form misses a closing parenthesis, but when I add the parenthesis then I get another error:
```
Next version 1.5.0-73ab116f
Loading Lisp file #P"/home/dalanicolai/.config/next/init.lisp"...

; file: /home/dalanicolai/.config/next/init.lisp
; in: DEFVAR *MY-KEYMAP*
;     (MAKE-KEYMAP "my-map")
; 
; caught STYLE-WARNING:
;   undefined function: MAKE-KEYMAP
; 
; compilation unit finished
;   Undefined function:
;     MAKE-KEYMAP
;   caught 1 STYLE-WARNING condition
While evaluating the form starting at line 1, column 0
  of #P"/home/dalanicolai/.config/next/init.lisp":
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {10006F05B3}>:
  Error: we could not load the init file:
The function COMMON-LISP-USER::MAKE-KEYMAP is undefined.

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10006F05B3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "~a:~&~a" {100638E853}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "~a:~&~a" {100638E853}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "~a:~&~a" {100638E853}>)
3: (UIOP/IMAGE:HANDLE-FATAL-CONDITION #<SIMPLE-ERROR "~a:~&~a" {100638E853}>)
4: (SB-KERNEL::%SIGNAL #<SIMPLE-ERROR "~a:~&~a" {100638E853}>)
5: (ERROR "~a:~&~a" "Error: we could not load the init file" #<UNDEFINED-FUNCTION MAKE-KEYMAP {100638E313}>)
6: (NEXT::LOAD-LISP #P"/home/dalanicolai/.config/next/init.lisp" :INTERACTIVE T)
7: (NEXT:START :URLS NIL :INIT-FILE NIL)
8: ((LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE))
9: (UIOP/IMAGE:CALL-WITH-FATAL-CONDITION-HANDLER #<CLOSURE (LAMBDA NIL :IN UIOP/IMAGE:RESTORE-IMAGE) {100621BE9B}>)
10: ((FLET SB-UNIX::BODY :IN SAVE-LISP-AND-DIE))
11: ((FLET "WITHOUT-INTERRUPTS-BODY-14" :IN SAVE-LISP-AND-DIE))
12: ((LABELS SB-IMPL::RESTART-LISP :IN SAVE-LISP-AND-DIE))

unhandled condition in --disable-debugger mode, quitting
```
Unfortunately as I do not know common-lisp (nor CLOS) yet, and because I don't know further introspection options yet, I can not provide much more/useful information...

**Comment by jmercouris** [comment]
Can you please post the entirety of your init file so that we may better assist you? Thanks!

**Comment by jmercouris** [comment]
Also you are correct, it should be:

```
(defvar *my-keymap* (make-keymap "my-map"))
```

**Comment by dalanicolai** [comment]
Hi thanks for looking into it. Because things were not working I removed all other code from the init file so that it contained the code I posted in my last comment above exclusively (where because of your last comment I now added that extra parenthesis). So the following is currently the full code in my init file:
```
(defvar *my-keymap* (make-keymap "my-map"))
  (define-key *my-keymap*
    "C-f" 'history-forwards
    "C-b" 'history-backwards)

  (define-mode my-mode ()
    "Dummy mode for the custom key bindings in `*my-keymap*'."
    ((keymap-scheme :initform (keymap:make-scheme
                               scheme:emacs *my-keymap*
                               scheme:vi-normal *my-keymap*))))

  (define-configuration buffer
      ((default-modes (append '(my-mode) %slot-default))))
```
When I try to run next then I get the error message (the second one because now I added the parenthesis) that I posted in my last comment above

**Comment by Ambrevar** [comment]
I've tried your snippet, it works for me.

From
From the output you've pasted, you're using commit 73ab116f which is 1
month old.  Try again against latest master, it should work.

**Comment by azzamsa** [comment]
@Ambrevar [This snippet](https://github.com/atlas-engineer/next/issues/447#issuecomment-562519272) doesn't work anymore.

I am getting this error:

``` bash
$ ./next
Next version 1.5.0
Loading configuration from #P"/home/john/.config/next/init.lisp"...

Error: we could not load the init file
READ error during LOAD:

  Package HOOKS does not exist.

    Line: 36, Column: 20

    Stream: #<SB-INT:FORM-TRACKING-STREAM for "file /home/john/.config/next/init.lisp" {1004C585F3}>
``` 

Any breaking changes or am I missing something?

**Comment by Ambrevar** [comment]
This topic is heavily outdated.  In the newest Next, run the `tutorial` command.  At the end you'll find a bunch of configuration snippets.

**Comment by Ambrevar** [comment]
Let me know if you need help with a particular config.

**Comment by dalanicolai** [comment]
@Ambrevar I wanted to try a few days ago with the newest version, but when I click the Download link in the git README then I got a "502 Bad Gateway" error, I assumed it would get fixed and today I wanted to try again but I still get the same error, despite #744 was closed yesterday because the Download page was/should be available again.

Also as recommended in #744, I tried to build from source but I get a build error. I saved the output of the make process in the following file:
[make_all.log](https://github.com/atlas-engineer/next/files/4664446/make_all.log)
Unfortunately I do not understand this log well yet. It seems to me that all the variables/flags are set correctly. (The command `make next` also errors)

**Comment by azzamsa** [comment]
> This topic is heavily outdated. In the newest Next, run the `tutorial` command. At the end you'll find a bunch of configuration snippets.

@Ambrevar Thanks for the tip. Unfortunately tutorial command only availiable at [2.0](https://github.com/atlas-engineer/next/blob/9ccda09c53676ae46349906b72af3041ac94228c/documents/CHANGELOG.org#L47). I still having problem [running next with guix from source](https://github.com/atlas-engineer/next/issues/148#issuecomment-632087248), Any hints?

**Comment by azzamsa** [comment]
I've tried to compile the tutorial to to html manually using `cl-markup`, this is the result.

![image](https://user-images.githubusercontent.com/17734314/82610165-1b3e3100-9be8-11ea-80a8-783a1eb4b35b.png)

I think we have typo in `(defvar *my-keymap*...`

I peak [your config](https://gitlab.com/ambrevar/dotfiles/-/blob/master/.config/next/init.lisp#L13) and covert mine to:

```lisp
(in-package :next-user)

(defvar *my-keymap* (make-keymap "my-map")
  "Keymap for `my-mode'.")

(define-key *my-keymap*
  "C-f" 'history-forwards
  "C-b" 'history-backwards)

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-scheme :initform (keymap:make-scheme
                             scheme:cua *my-keymap*
                             scheme:emacs *my-keymap*
                             scheme:vi-normal *my-keymap*))))

(define-configuration buffer
  ((default-modes (append '(my-mode) %slot-default))))
```

but still have the same error:

``` bash
$ ./next
Next version 1.5.0
Loading configuration from #P"/home/user/.config/next/init.lisp"...
While evaluating the form starting at line 3, column 0
  of #P"/home/user/.config/next/init.lisp":
Error: we could not load the init file
invalid number of arguments: 1
```

**Comment by Ambrevar** [comment]
Daniel Nicolai <notifications@github.com> writes:

> @Ambrevar I wanted to try a few days ago with the newest version, but when I click the Download link in the git README then I got a "502 Bad Gateway" error, I assumed it would get fixed and today I wanted to try again but I still get the same error, despite #744 was closed yesterday because the Download page was/should be available again.

Which link?
https://next.atlas.engineer/download is working for me, but if you want
the newest version you should go for master on GitHub for now.

> Also as recommended in #744, I tried to build from source but I get a build error. I saved the output of the make process in the following file:
> [make_all.log](https://github.com/atlas-engineer/next/files/4664446/make_all.log)

You are probably using an outdated Quicklisp.
Can you open a new issue and detail how you've built Next?
(Commands you've entered and the results you got.)

**Comment by Ambrevar** [comment]
@azzamsa: Good catch!

I've fixed the typo in 4ba0a9384cbfee330069ab4052d64c1fb865c894.

About your error: the new config style only works on master.

If you want to consult documentation for 1.5, have a look at

https://github.com/atlas-engineer/next/blob/1.5.0/documents/MANUAL.org

**Comment by dalanicolai** [comment]
@Ambrevar I indeed meant that download link which was not working when I created my last comment (and also the previous time), but it is working now here too.

Actually to get build instructions for next, I tried to access the developer manual page but that also gave error 502 at that time. Therefore I followed the instructions in the INSTALL file in the repo. However, I had two copies of the repo, one in my ~/git directory and another one in my ~/common-lisp directory and I had forgotten about the latter one. So I was following the instruction in the repo in my ~/git directory. Now that I could access the Developer Manual again, I found that I should build from the repo in my ~/common-lisp directory. So that solves the issue of my last comment.
But I suggest to also mention in the INSTALL file that the build should be done from a repo within the ~/common-lisp directory.

However, some build error remains even after pulling the latest version of the repo. So as you requested I will open a new issue for that.

**Comment by Ambrevar** [comment]
The INSTALL file is the step after checking out.

The instructions for checking out the repository are in
documents/README.org, which mentions the ~/common-lisp folder.

Does that make sense?

**Comment by dalanicolai** [comment]
Well it sounds sensible but maybe it makes more sense when I get more
experienced. For now I am happy if "full" build instructions, or references
to these instructions are in the INSTALL file. Also, I've build next
successfully before, and it would be great to not have to read the README
again if I want to rebuild it again. In my opinion it would be worth it to
remind in the INSTALL file about updating to latest versions of quicklisp
and cl-webkit (and next repo itself), and additionally that for successfull
completion of the INSTALL file instructions it should be located in the
`~/common-lisp/next` directory, in that way it is easier for ordinary
users, i.e. non-developers, like me to contribute useful feedback.

**Comment by Ambrevar** [comment]
I've added a warning about outdated quickdist / libraries in
0d3335d78ad6e6104bd37a15982e6f2fe7af02e2.

Note that Next can be cloned outside of ~/common-lisp.  This
recommendation is mostly for people who want to load Next from a Common
Lisp REPL (e.g. SLIME / SLY).

The INSTALL instructions are meant to be quite stable.  The reason it
didn't work out for you previously is because you were using an outdated
Next checkout that didn't automatically update the Quickdist /
cl-webkit.  We do this now.

Let me know if there is anything else.

**Comment by azzamsa** [comment]
@Ambrevar I've successfully installed latest next with guix. However the `customizing keybinding` example in docs doesn't work. I've convert my old config but it also doesn't work:

``` lisp
(in-package :next-user)

(defvar *my-keymap* (make-keymap "my-map"))

(define-key *my-keymap*
  "C-m" 'switch-buffer)

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-scheme :initform (keymap:make-scheme
                             scheme:emacs *my-keymap*
                             scheme:vi-normal *my-keymap*))))

(define-configuration buffer
  ((default-modes (append '(my-mode) %slot-default))))
```

And I can't do `M-x` to call any function.

**Comment by Ambrevar** [comment]
Are you using CUA, Emacs or VI binding?

The default scheme is CUA.  Bring up the command menu with C-space.
This should be explained (with dynamic bindings) on the start page.

In your example you haven't set the keymap for scheme:cua which is why
you may not be able to call it from the CUA scheme.

**Comment by azzamsa** [comment]
this works for me

``` lisp
(in-package :next-user)

(defvar *my-keymap* (make-keymap "my-map"))

(define-key *my-keymap*
  "S-m" 'switch-buffer)

(define-mode my-mode ()
  "Dummy mode for the custom key bindings in `*my-keymap*'."
  ((keymap-scheme :initform (keymap:make-scheme
                             scheme:cua *my-keymap*
                             scheme:emacs *my-keymap*
                             scheme:vi-normal *my-keymap*))))


(define-configuration buffer
  ((default-modes (append '(my-mode emacs-mode) %slot-default))))

```

**Comment by Ambrevar** [comment]
> (in-package :next-user)

You don't need this line as it is the default.

> (defvar *my-keymap* (make-keymap "my-map"))
>
> (define-key *my-keymap*
>   "S-m" 'switch-buffer)
>
> (define-mode my-mode ()
>   "Dummy mode for the custom key bindings in `*my-keymap*'."
>   ((keymap-scheme :initform (keymap:make-scheme
>                              scheme:cua *my-keymap*
>                              scheme:emacs *my-keymap*
>                              scheme:vi-normal *my-keymap*))))
>
>
> (define-configuration buffer
>   ((default-modes (append '(my-mode emacs-mode) %slot-default))))

This is what is presented in the tutorial (minus the emacs-mode
default).
Would you like to suggest an improvement to the documentation?

**Comment by azzamsa** [comment]
> Would you like to suggest an improvement to the documentation?

Sure!

**Comment by Ambrevar** [comment]
Which one?
